File: NetSim.py
import json

try:
    from ns import ns
except ModuleNotFoundError:
    raise SystemExit(
        "Error: ns3 Python module not found;"
        " Python bindings may not be enabled"
        " or your PYTHONPATH might not be properly configured"
    )

cmd = ns.CommandLine(__file__)

"""
Structure of the Simulated network should look somewhat like this:


    Prosumer 1  <--Point-to-Point--> Bridge 1 <--CSMA--> Bridge 2 <--Point-to-Point--> Grid Operator
    Prosumer 2  <--Point-to-Point--> Bridge 1 <--CSMA--> Bridge 2 <--Point-to-Point--> Grid Operator
    
    
    TODO: Müsste aber vielleicht eher so Aussehen:
    
    Prosumer 1  <--CSMA--> 
                            Bridge 1 <--Point-to-Point--> Bridge 2 <--Point-to-Point--> Grid Operator
    Prosumer 2  <--CSMA--> 
    
etc.
Basically, there is the bridge connection, which on the one side is connected to all the prosumers 
and on the other side to the grid operator.
    
"""


def simulate_SGMW_TAF14():
    # Set Data Rate and Packet Size for the OnOffApplication
    ns.Config.SetDefault("ns3::OnOffApplication::PacketSize", ns.StringValue("1024"))
    ns.Config.SetDefault("ns3::OnOffApplication::DataRate", ns.StringValue("5Mb/s"))

    # Create the prosumer nodes
    prosumer_nodes = ns.NodeContainer()
    prosumer_nodes.Create(2)

    # Create some bridge nodes
    bridge_nodes = ns.NodeContainer()
    bridge_nodes.Create(2)

    # Create the CSMA devices for the bridge nodes
    csma = ns.CsmaHelper()
    csma.SetChannelAttribute("DataRate", ns.StringValue("100Mbps"))
    csma.SetChannelAttribute("Delay", ns.StringValue("2ms"))
    bridge_devices = csma.Install(bridge_nodes)

    # Create the grid operator node
    grid_operator_node = ns.NodeContainer()
    grid_operator_node.Create(1)

    # Create the point-to-point link between the prosumer and the bridge nodes and the bridge and grid operator
    point_to_point = ns.PointToPointHelper()
    point_to_point.SetDeviceAttribute("DataRate", ns.StringValue("5Mbps"))
    point_to_point.SetChannelAttribute("Delay", ns.StringValue("2ms"))

    prosumer_devices = []

    # Create the point-to-point link between the prosumer and the first bridge node
    for i in range(0, prosumer_nodes.GetN()):
        device = point_to_point.Install(prosumer_nodes.Get(i), bridge_nodes.Get(0))
        prosumer_devices.append(device)

    # Create the point-to-point link between the bridge nodes and the grid operator
    grid_operator_device = point_to_point.Install(bridge_nodes.Get(1), grid_operator_node.Get(0))

    tap_bridge = ns.TapBridgeHelper()
    # tap_bridge.SetAttribute("Mode", ns.StringValue("ConfigureLocal"))
    # tap_bridge.Install(grid_operator_node.Get(0), "ns3")

    # Install the internet stack on the prosumer and grid operator nodes
    stack = ns.InternetStackHelper()
    olsr = ns.OlsrHelper()
    # Somehow this stops the simulation from continuing
    stack.SetRoutingHelper(olsr)
    stack.InstallAll()

    # Declare the IP address space
    address = ns.Ipv4AddressHelper()
    address.SetBase(ns.Ipv4Address("10.1.1.0"), ns.Ipv4Mask("255.255.255.0"))

    # Assign IP addresses to the bridge nodes and the grid operator node
    bridge_addresses = address.Assign(bridge_devices)
    grid_operator_address = address.Assign(grid_operator_device)

    interfaces = []
    # Assign IP addresses to the prosumer and bridge nodes
    for i in range(len(prosumer_devices)):
        interface = address.Assign(prosumer_devices[i])
        interfaces.append(interface)

    # Create an address list -> Addresses are saved in order of interfaces
    address_list = []
    for i in range(len(interfaces)):
        address_list.append(interfaces[i].GetAddress(0))
        address_list.append(interfaces[i].GetAddress(1))

    # Create the Server Application on the grid operator node
    port = 8080
    server_helper = ns.PacketSinkHelper("ns3::TcpSocketFactory",
                                        ns.InetSocketAddress(grid_operator_address.GetAddress(1), port).ConvertTo())
    server_apps = server_helper.Install(grid_operator_node.Get(0))
    grid_operator_node.Get(0).AddApplication(server_apps.Get(0))
    point_to_point.EnablePcap("GO", grid_operator_node.Get(0), True)
    server_apps.Start(ns.Seconds(1.0))
    server_apps.Stop(ns.Seconds(10.0))

    # Create a TCP client on the first prosumer node
    client_helper = ns.OnOffHelper("ns3::TcpSocketFactory", ns.InetSocketAddress(address_list[0], port).ConvertTo())
    client_helper.SetAttribute("MaxBytes", ns.UintegerValue(1024))
    client_apps = client_helper.Install(prosumer_nodes.Get(0))
    prosumer_nodes.Get(0).AddApplication(client_apps.Get(0))
    point_to_point.EnablePcap("Pro", prosumer_nodes.Get(0), True)

    client_apps.Start(ns.Seconds(2.0))
    client_apps.Stop(ns.Seconds(10.0))

    # tcp_socket = ns.TcpSocketBase()
    # tcp_socket.SetNode(prosumer_nodes.Get(1))

    # The next two throw null pointer exceptions
    # tcp_socket.Bind(ns.InetSocketAddress(address_list[2], port).ConvertTo())
    # tcp_socket.Connect(ns.InetSocketAddress(address_list[3], port).ConvertTo())

    """ not really working (or necessary?) at the moment
    # Probably unnecessary code for the packet socket because of TCP socket
    #  -> throws errors at the moment
    # packet_helper = ns.PacketSocketHelper()
    # packet_app = packet_helper.Install(prosumer_nodes.Get(1))
    # packet_socket = packet_app.CreateSocket()
    # packet_socket.Bind(ns.InetSocketAddress(interfaces[1].GetAddress(0), port).ConvertTo())
    """

    # Create a packet -> will be replaced by the measurement data later
    data = "TestPacket"
    data_e = bytearray(data.encode())
    packet = ns.Packet(data_e, 1024)

    # tcp_socket.SendTo(packet, 0, ns.InetSocketAddress(address_list[3], port).ConvertTo())
    # tcp_socket.Send(packet, 0)
    # tcp_socket.Close()

    # Maybe this is also a valid way to send a packet -> last one is protocol number
    # -> TCP should be 6 according to google, UDP is 17
    # print(prosumer_devices[1].Get(0).GetAddress())
    # prosumer_devices[1].Get(0).SendFrom(packet, ns.InetSocketAddress(address_list[2], port).ConvertTo(),
    #                          ns.InetSocketAddress(grid_operator_address.GetAddress(1), port).ConvertTo(), 6)
    # print(devices[1].Get(1).Receive(packet))
    # Still nothing is received
    # print(server_apps.Get(0).GetTotalRx())

    # point_to_point.EnablePcapAll("SGMW_TAF14")

    ns.Simulator.Stop(ns.Seconds(11.0))
    ns.Simulator.Run()
    ns.Simulator.Destroy()

    print("Simulation has ended")
if __name__ == "__main__":
    simulate_SGMW_TAF14()

File: NetSim.cc
#include "ns3/core-module.h"
#include "ns3/network-module.h"
#include "ns3/internet-module.h"
#include "ns3/point-to-point-module.h"
#include "ns3/applications-module.h"
#include "ns3/pcap-file-wrapper.h"

#include <nlohmann/json.hpp>

#include "NetSimApp.hpp"
#include "NetSimSinkApp.hpp"
#include "helpers.hpp"

#include <fstream>
#include <sstream>

using namespace ns3;
using json = nlohmann::json;

NS_LOG_COMPONENT_DEFINE("TcpExample");

void SendPacketTrace(Ptr<const Packet> packet)
{
    NS_LOG_UNCOND("Packet Sent: " + std::to_string(packet->GetSize()));
}

void ReceivePacketTrace(Ptr<const Packet> packet, const Address& from, const Address& to)
{
    Ipv4Address ipv4AddrFrom = InetSocketAddress::ConvertFrom(from).GetIpv4();
    Ipv4Address ipv4AddrTo = InetSocketAddress::ConvertFrom(to).GetIpv4();
    std::ostringstream oss;
    oss << "Packet Received from: " << ipv4AddrFrom << " to: " << ipv4AddrTo;
    NS_LOG_UNCOND(oss.str());

    // Extract the payload from the packet and convert it to a string
    std::vector<uint8_t> buffer(packet->GetSize());
    packet->CopyData(buffer.data(), buffer.size());
    std::string jsonData(buffer.begin(), buffer.end());

    // Parse the string as JSON and save it to a file
    try {
        json receivedJson = json::parse(jsonData);

        // Open "grid_data.json" in append mode to add the received JSON data
        std::ofstream outFile("JSON/grid_data.json", std::ios::app);

        // Write the JSON data to the file
        outFile << receivedJson.dump() << std::endl;
    } catch (json::parse_error& e) {
        NS_LOG_UNCOND("Failed to parse JSON: " + std::string(e.what()));
    }
}

void simulation_loop(uint32_t n_nodes) {

  // Receive `n_nodes` amount of messages from the GridSim.py for further
  // processing
  auto measurements = receive_messages(n_nodes, 8081);

  NetSimApp appInstance;

  NodeContainer nodes;
  nodes.Create(n_nodes +
               1); // create a node for every measurement + one for the operator

  PointToPointHelper pointToPoint;
  pointToPoint.SetDeviceAttribute("DataRate", StringValue("5Mbps"));
  pointToPoint.SetChannelAttribute("Delay", StringValue("2ms"));

  InternetStackHelper stack;
  stack.Install(nodes);

  Ipv4AddressHelper address;
  std::vector<Ipv4InterfaceContainer> interfaces(n_nodes);

  // connect all nodes to the operator node
  for (uint32_t i = 0; i < n_nodes; ++i) {
    NetDeviceContainer devices =
        pointToPoint.Install(NodeContainer(nodes.Get(i), nodes.Get(n_nodes)));
    address.SetBase(std::format("10.0.{}.0", i + 1).c_str(), "255.255.255.0");
    interfaces[i] = address.Assign(devices);
  }

  // install a packet sink on the operator node for every sender
  uint16_t port = 8080;
  std::vector<Ptr<Socket>> sockets;
  for (int i = 0; i < n_nodes; ++i) {
    Ptr<Socket> sinkSocket =
        Socket::CreateSocket(nodes.Get(n_nodes), TcpSocketFactory::GetTypeId());
    sinkSocket->Bind(InetSocketAddress(interfaces[i].GetAddress(1), 8080));
    sockets.push_back(sinkSocket);
  }
  Ptr<SinkApp> sinkApp = Create<SinkApp>(sockets);
  nodes.Get(n_nodes)->AddApplication(sinkApp);
  sinkApp->SetStartTime(Seconds(0.5));
  sinkApp->SetStopTime(Seconds(20.));

  PacketSinkHelper packetSinkHelper(
      "ns3::TcpSocketFactory", InetSocketAddress(Ipv4Address::GetAny(), port));
  ApplicationContainer sinkApps = packetSinkHelper.Install(nodes.Get(n_nodes));
  sinkApps.Start(Seconds(1.0));
  sinkApps.Stop(Seconds(10.0));

  for (uint32_t i = 0; i < n_nodes; ++i) {
    Ptr<Socket> ns3TcpSocket =
        Socket::CreateSocket(nodes.Get(i), TcpSocketFactory::GetTypeId());
    Address sinkAddress(InetSocketAddress(interfaces[i].GetAddress(1), port));
    Ptr<NetSimApp> app = CreateObject<NetSimApp>();
    app->Setup(ns3TcpSocket, sinkAddress, 1040, DataRate("1Mbps"),
               measurements[i]);
    nodes.Get(i)->AddApplication(app);
    app->SetStartTime(Seconds(2.0));
    app->SetStopTime(Seconds(10.0));

    app->TraceConnectWithoutContext("Tx", MakeCallback(&SendPacketTrace));
  }

  nodes.Get(n_nodes)->TraceConnectWithoutContext(
      "RxWithAddresses", MakeCallback(&ReceivePacketTrace));

  //   Config::ConnectWithoutContext(
  //       std::format(
  //           "/NodeList/{}/ApplicationList/*/$ns3::PacketSink/RxWithAddresses",
  //           n_nodes)
  //           .c_str(),
  //       MakeCallback(&ReceivePacketTrace));

  pointToPoint.EnablePcapAll("PCAP/NetSim");

  Simulator::Run();
  Simulator::Destroy();

  auto recvd_measurements = sinkApp->getMeasurements();
  auto aggregated_measurements = collect_jsons(recvd_measurements);

  // Send the JSON data to the Python simulation via netcat
  network::send_message("127.0.0.1", 10000, aggregated_measurements.dump());
}

int main(int argc, char* argv[])
{
  // receive the number of simulation nodes nodes
  auto n_nodes_str = network::wait_for_message(10000);
  auto n_nodes = std::stoi(std::string{n_nodes_str.begin(), n_nodes_str.end()});

  // keep reading in measurements, running the netsim and responding with the
  // aggregated results
  LogComponentEnable("TcpExample", LOG_LEVEL_INFO);
  while (true) {
    simulation_loop(n_nodes);
  }
    return 0;
}